docs/source/utils/generate_examples.py

```diff
@@ -0,0 +1,134 @@
+            contents += generate_nb_gallery(f"{included[0]}.{subpackage[0]}", files)
+
+    destination.parent.mkdir(exist_ok=True, parents=True)
+    destination.write_text(contents)
+
+
+def sort_example_file_tree(files: Set[Path]) -> List[Path]:
+    """
+    Sort files alphabetically; for the example files (whose names start with number) numerical sort is applied.
+
+    :param files: Files list to sort.
+    """
+    examples = {file for file in files if file.stem.split("_")[0].isdigit()}
+    return sorted(examples, key=lambda file: int(file.stem.split("_")[0])) + sorted(files - examples)
+
+
+def iterate_examples_dir_generating_links(source: Path, dest: Path, base: str) -> List[Path]:
+    """
+    Recursively travel through examples directory, creating links for all files under docs/source/examples/ root.
+    Created link files have dot-path name matching source file tree structure.
+
+    :param source: Examples root (usually examples/).
+    :param dest: Examples destination (usually docs/source/examples/).
+    :param base: Dot path to current dir (will be used for link file naming).
+    """
+    if not source.is_dir():
+        raise Exception(f"Entity {source} appeared to be a file during processing!")
+    links = list()
+    for entity in [obj for obj in sort_example_file_tree(set(source.glob("./*"))) if not obj.name.startswith("__")]:
+        base_name = f"{base}.{entity.name}"
+        if entity.is_file() and entity.suffix in (".py", ".ipynb"):
+            base_path = Path(base_name)
+            create_notebook_link(entity, dest / base_path)
+            links += [base_path]
+        elif entity.is_dir() and not entity.name.startswith("_"):
+            links += iterate_examples_dir_generating_links(entity, dest, base_name)
+    return links
+
+
+def generate_example_links_for_notebook_creation(
+    include: Optional[List[Union[Tuple[str, str], Tuple[str, List[Tuple[str, str]]]]]] = None,
+    exclude: Optional[List[str]] = None,
+    source: str = "examples",
+    destination: str = "docs/source/examples",
+):
+    """
+    Generate symbolic links to examples files (examples/) in docs directory (docs/source/examples/).
+    That is required because Sphinx doesn't allow to include files from parent directories into documentation.
+    Also, this function creates index files inside each generated folder.
+    That index includes each folder contents, so any folder can be imported with 'folder/index'.
+
+    :param include: Files to copy (supports file templates, like *).
+    :param exclude: Files to skip (supports file templates, like *).
+    :param source: Examples root, default: 'examples/'.
+    :param destination: Destination root, default: 'docs/source/examples/'.
+    """
+    include = [("examples", "Examples")] if include is None else include
+    exclude = list() if exclude is None else exclude
+    dest = Path(destination)
+
+    flattened = list()
+    for package in include:
+        if len(package) == 2:
```

Comment:
Again, isn't it always of that len?

`package` is of type `Union[Tuple[str, str], Tuple[str, List[Tuple[str, str]]]]`

Reply:
@pseusys